File: verification/scripts/python/regularized/test_ridge.py
# ============================================================================
# Ridge Regression Test Reference Implementation (Python with sklearn)
# ============================================================================
# This script generates reference values for ridge regression using Python's
# sklearn library. The test validates that the Rust implementation matches
# sklearn's Ridge behavior for coefficients and predictions.
#
# Source: sklearn.linear_model.Ridge
# Reference: "Regularization Paths for Generalized Linear Models via
#            Coordinate Descent" (Friedman, Hastie, Tibshirani 2010)
#
# Usage:
#   python test_ridge.py [csv_path] [output_dir] [lambda_count]
#   Args:
#     csv_path    - Path to CSV file (first col = response, rest = predictors)
#                   Default: ../../datasets/csv/mtcars.csv
#     output_dir  - Path to output directory
#                   Default: ../../results/python
#     lambda_count- Number of lambda values to generate
#                   Default: 20
# ============================================================================

import numpy as np
import pandas as pd
from sklearn.linear_model import Ridge
from sklearn.preprocessing import StandardScaler
import json
import os
import sys
from pathlib import Path

# Helper for JSON serialization of numpy types
class NumpyEncoder(json.JSONEncoder):
    def default(self, obj):
        if isinstance(obj, np.integer):
            return int(obj)
        if isinstance(obj, np.floating):
            return float(obj)
        if isinstance(obj, np.ndarray):
            return obj.tolist()
        return super(NumpyEncoder, self).default(obj)


def generate_lambda_sequence(y, X, n_lambdas=20):
    """
    Generate a lambda sequence similar to glmnet's approach.
    glmnet starts from lambda_max (the smallest lambda where all coefficients
    are zero) and decreases logarithmically.
    """
    n, p = X.shape

    # Compute lambda_max: smallest lambda where all coefficients would be zero
    # For ridge, this is theoretically infinity, but we use a practical upper bound
    # based on the OLS solution scaled up
    from sklearn.linear_model import LinearRegression
    ols = LinearRegression(fit_intercept=True)
    ols.fit(X, y)

    # Use cross-product of X and y as reference
    X_centered = X - X.mean(axis=0)
    y_centered = y - y.mean()

    # Maximum lambda that gives meaningful shrinkage
    lambda_max = np.max(np.abs(X_centered.T @ y_centered)) / n * 10

    # Logarithmic sequence from lambda_max to lambda_max * 1e-4
    lambda_sequence = np.logspace(np.log10(lambda_max), np.log10(lambda_max * 1e-4), n_lambdas)

    return lambda_sequence[::-1]  # Descending order like glmnet


def convert_categorical_to_numeric(data, dataset_name):
    """Convert categorical columns to numeric representations."""
    non_numeric_cols = data.select_dtypes(exclude=[np.number]).columns.tolist()

    if len(non_numeric_cols) > 0:
        print(f"INFO: Dataset '{dataset_name}' contains non-numeric columns: "
              f"{', '.join(non_numeric_cols)}")
        print("Converting categorical variables to numeric representations...")

        for col in non_numeric_cols:
            data[col] = pd.to_numeric(data[col], errors='coerce')
            # Fill NaN with mode
            if data[col].isna().any():
                mode_val = data[col].mode()
                if len(mode_val) > 0:
                    data[col].fillna(mode_val[0], inplace=True)

    return data


def main():
    # Parse command line arguments
    args = sys.argv[1:]

    # Set defaults
    default_csv = "../../datasets/csv/mtcars.csv"
    default_output = "../../results/python"
    default_lambda_count = 20

    csv_path_input = args[0] if len(args) >= 1 else default_csv
    output_dir_input = args[1] if len(args) >= 2 else default_output
    lambda_count = int(args[2]) if len(args) >= 3 else default_lambda_count

    # Resolve paths - use absolute paths directly, resolve relative paths from cwd
    csv_path = Path(csv_path_input) if Path(csv_path_input).is_absolute() else Path.cwd() / csv_path_input
    output_dir = Path(output_dir_input) if Path(output_dir_input).is_absolute() else Path.cwd() / output_dir_input

    # Validate CSV path
    if not csv_path.exists():
        print(f"ERROR: CSV file not found: {csv_path}")
        sys.exit(1)

    # Extract dataset name from filename
    dataset_name = csv_path.stem

    print(f"Running ridge regression test on dataset: {dataset_name}")

    # Load data
    data = pd.read_csv(csv_path)
    data = convert_categorical_to_numeric(data, dataset_name)

    # Extract response (first column) and predictors (remaining columns)
    y = data.iloc[:, 0].values
    X = data.iloc[:, 1:].values

    n, p = X.shape

    print(f"  n = {n} observations, p = {p} predictors")

    # Generate lambda sequence
    lambda_sequence = generate_lambda_sequence(y, X, lambda_count)

    # Store results for each lambda
    coefficients_list = []
    degrees_of_freedom = []

    # Standardize X for computing effective degrees of freedom
    scaler_X = StandardScaler()
    X_standardized = scaler_X.fit_transform(X)

    for lam in lambda_sequence:
        # Fit ridge regression
        # sklearn uses alpha = lambda / n (approximately)
        # We need to match glmnet's objective function
        # glmnet: (1/2n) * ||y - Xb||^2 + (lambda/2) * ||b||^2
        # sklearn: ||y - Xb||^2 + alpha * ||b||^2
        # So alpha = lambda
        ridge = Ridge(alpha=lam, fit_intercept=True,
                      solver='auto', random_state=42)

        # sklearn expects standardized data internally for Ridge
        # We'll manually standardize to match glmnet's behavior
        X_mean = X.mean(axis=0)
        X_std = X.std(axis=0, ddof=0)  # ddof=0 for population std (glmnet uses this)

        # Handle constant columns
        X_std_safe = np.where(X_std < 1e-10, 1.0, X_std)

        X_scaled = (X - X_mean) / X_std_safe

        # Fit on standardized data
        ridge.fit(X_scaled, y)

        # Get coefficients on original scale
        # coef_original = coef_scaled / std
        coef_scaled = ridge.coef_
        intercept = ridge.intercept_ - np.sum(coef_scaled * X_mean / X_std_safe)
        coef_original = coef_scaled / X_std_safe

        # Store coefficients (intercept first, then coefficients)
        coefs_with_intercept = np.concatenate([[intercept], coef_original])
        coefficients_list.append(coefs_with_intercept.tolist())

        # Compute effective degrees of freedom for ridge
        # df = trace(X (X'X + lambda*I)^-1 X')
        # For standardized X: df = sum(s_i^2 / (s_i^2 + lambda)) where s_i are singular values
        U, s, Vt = np.linalg.svd(X_standardized, full_matrices=False)
        df_lam = np.sum(s**2 / (s**2 + lam))
        degrees_of_freedom.append(df_lam)

    # Get predictions at a few representative lambdas
    test_indices = [0, lambda_count // 2, lambda_count - 1]
    test_lambdas = lambda_sequence[test_indices].tolist()

    # Create test data (use first min(5, n) rows for prediction tests)
    n_test = min(5, n)
    X_test = X[:n_test, :]

    predictions = []
    for idx in test_indices:
        lam = lambda_sequence[idx]

        ridge = Ridge(alpha=lam, fit_intercept=True, solver='auto', random_state=42)
        X_mean = X.mean(axis=0)
        X_std = X.std(axis=0, ddof=0)
        X_std_safe = np.where(X_std < 1e-10, 1.0, X_std)
        X_scaled = (X - X_mean) / X_std_safe
        ridge.fit(X_scaled, y)

        # Predict on test data
        X_test_scaled = (X_test - X_mean) / X_std_safe
        pred = ridge.predict(X_test_scaled)
        predictions.append(pred.tolist())

    # Compute fitted values at the final lambda
    lam_final = lambda_sequence[-1]
    ridge_final = Ridge(alpha=lam_final, fit_intercept=True, solver='auto', random_state=42)
    X_mean = X.mean(axis=0)
    X_std = X.std(axis=0, ddof=0)
    X_std_safe = np.where(X_std < 1e-10, 1.0, X_std)
    X_scaled = (X - X_mean) / X_std_safe
    ridge_final.fit(X_scaled, y)

    fitted_values = ridge_final.predict(X_scaled).tolist()
    residuals = (y - np.array(fitted_values)).tolist()

    # Build result object
    result = {
        "test": "ridge",
        "method": "sklearn",
        "alpha": 0,
        "n": n,
        "p": p,
        "lambda_sequence": lambda_sequence.tolist(),
        "coefficients": coefficients_list,
        "degrees_of_freedom": degrees_of_freedom,
        "test_lambdas": test_lambdas,
        "test_predictions": predictions,
        "fitted_values": fitted_values,
        "residuals": residuals,
        "sklearn_version": sklearn.__version__
    }

    # Ensure output directory exists
    output_dir.mkdir(parents=True, exist_ok=True)

    # Write output
    output_file = output_dir / f"{dataset_name}_ridge.json"
    with open(output_file, 'w') as f:
        json.dump(result, f, cls=NumpyEncoder, indent=2)

    print(f"Wrote: {output_file.absolute()}")


if __name__ == "__main__":
    import sklearn  # Import here for version check
    main()

File: verification/scripts/r/regularized/test_ridge.R
# ============================================================================
# Ridge Regression Test Reference Implementation (R with glmnet)
# ============================================================================
# This script generates reference values for ridge regression using R's
# glmnet package with alpha=0. The test validates that the Rust implementation
# matches glmnet's behavior for coefficient paths and predictions.
#
# Source: glmnet package, glmnet function with alpha=0
# Reference: Friedman, Hastie, Tibshirani (2010), "Regularization Paths for
#            Generalized Linear Models via Coordinate Descent"
#
# Usage:
#   Rscript test_ridge.R [csv_path] [output_dir] [lambda_count]
#   Args:
#     csv_path    - Path to CSV file (first col = response, rest = predictors)
#                   Default: ../../../datasets/csv/mtcars.csv
#     output_dir  - Path to output directory
#                   Default: ../../../results/r
#     lambda_count- Number of lambda values to generate
#                   Default: 20
# ============================================================================

# Get script directory and resolve relative paths
# Get the script path from commandArgs (works with Rscript)
args_all <- commandArgs(trailingOnly = FALSE)
script_arg <- args_all[grep("^--file=", args_all)]
if (length(script_arg) > 0) {
  script_path <- dirname(sub("--file=", "", script_arg))
  original_wd <- getwd()
  setwd(script_path)
} else {
  original_wd <- getwd()
}

# Add user library to path
user_lib <- file.path(Sys.getenv("USERPROFILE"), "Documents", "R", "win-library", "4.4")
if (dir.exists(user_lib)) {
    .libPaths(c(user_lib, .libPaths()))
}

# Load required packages
if (!require("glmnet", quietly = TRUE)) {
    stop("Package 'glmnet' is required. Install with: install.packages('glmnet')")
}
if (!require("jsonlite", quietly = TRUE)) {
    stop("Package 'jsonlite' is required. Install with: install.packages('jsonlite')")
}

# Helper function to convert categorical columns to numeric
convert_categorical_to_numeric <- function(data, dataset_name) {
  non_numeric_cols <- names(data)[sapply(data, function(x) !is.numeric(x))]

  if (length(non_numeric_cols) > 0) {
    cat(paste0("INFO: Dataset '", dataset_name, "' contains non-numeric columns: ",
               paste(non_numeric_cols, collapse = ", "), "\n"))
    cat("Converting categorical variables to numeric representations...\n")

    for (col in non_numeric_cols) {
      if (is.factor(data[[col]])) {
        data[[col]] <- as.numeric(data[[col]])
      } else if (is.character(data[[col]])) {
        temp_numeric <- as.numeric(data[[col]])
        if (any(is.na(temp_numeric))) {
          data[[col]] <- as.numeric(as.factor(data[[col]]))
        } else {
          data[[col]] <- temp_numeric
        }
      }
    }
  }

  return(data)
}

# Parse command line arguments
args <- commandArgs(trailingOnly = TRUE)

# Set defaults (relative to original working directory)
default_csv <- "../../../datasets/csv/mtcars.csv"
default_output <- "../../../results/r"
default_lambda_count <- 20

# Resolve paths relative to original working directory (before setwd)
resolve_path <- function(path) {
  # Check if path is absolute (starts with / on Unix, or drive letter on Windows)
  is_absolute <- grepl("^(/|[A-Za-z]:)", path)
  if (!is_absolute) {
    # Relative path - resolve from original working directory
    file.path(original_wd, path)
  } else {
    path
  }
}

# Parse arguments
csv_path_raw <- ifelse(length(args) >= 1, args[1], default_csv)
output_dir_raw <- ifelse(length(args) >= 2, args[2], default_output)
lambda_count <- as.integer(ifelse(length(args) >= 3, args[3], default_lambda_count))

# Resolve paths
csv_path <- resolve_path(csv_path_raw)
output_dir <- resolve_path(output_dir_raw)

# Validate CSV path
if (!file.exists(csv_path)) {
  stop(paste("CSV file not found:", csv_path))
}

# Extract dataset name from filename
dataset_name <- tools::file_path_sans_ext(basename(csv_path))

cat(sprintf("Running ridge regression test on dataset: %s\n", dataset_name))

# Load data
data <- read.csv(csv_path)
data <- convert_categorical_to_numeric(data, dataset_name)

# Extract response (first column) and predictors (remaining columns)
y <- data[, 1]
X <- as.matrix(data[, -1, drop = FALSE])

n <- nrow(X)
p <- ncol(X)

cat(sprintf("  n = %d observations, p = %d predictors\n", n, p))

# Run glmnet for ridge regression (alpha = 0)
set.seed(42)  # For reproducibility
fit <- glmnet(X, y, family = "gaussian", alpha = 0,
             standardize = TRUE, intercept = TRUE,
             nlambda = lambda_count)

# Extract coefficient matrix (includes intercept in first row)
coef_matrix <- as.matrix(coef(fit))
# coef_matrix is (p+1) x lambda_count, with intercept in row 1

# Convert to list format for JSON serialization
# Each lambda gets a list of coefficients (including intercept as first element)
coef_list <- split(coef_matrix, rep(1:ncol(coef_matrix), each = nrow(coef_matrix)))
coef_list <- lapply(coef_list, as.vector)

# Get lambda sequence
lambda_seq <- fit$lambda

# Get degrees of freedom for each lambda
df <- fit$df

# Get predictions at a few representative lambdas
# Use actual length of lambda sequence (glmnet may generate fewer than requested)
actual_lambda_count <- length(lambda_seq)
test_indices <- c(1, ceiling(actual_lambda_count/2), actual_lambda_count)
test_lambdas <- lambda_seq[test_indices]

# Create test data (use first min(5, n) rows for prediction tests)
n_test <- min(5, n)
X_test <- X[1:n_test, , drop = FALSE]

predictions <- list()
for (i in test_indices) {
  pred <- predict(fit, newx = X_test, s = lambda_seq[i])
  predictions[[length(predictions) + 1]] <- as.vector(pred)
}

# Compute fitted values at the final lambda
fitted_values <- as.vector(predict(fit, newx = X, s = lambda_seq[actual_lambda_count]))

# Compute residuals at the final lambda
residuals <- as.vector(y - fitted_values)

# Build result object
result <- list(
  test = "ridge",
  method = "glmnet",
  alpha = 0,
  n = n,
  p = p,
  lambda_sequence = as.vector(lambda_seq),
  coefficients = coef_list,
  degrees_of_freedom = as.vector(df),
  test_lambdas = as.vector(test_lambdas),
  test_predictions = predictions,
  fitted_values = fitted_values,
  residuals = residuals,
  glmnet_version = as.character(packageVersion("glmnet"))
)

# Write output
output_file <- file.path(output_dir, paste0(dataset_name, "_ridge.json"))
write_json(result, output_file, pretty = TRUE, auto_unbox = TRUE)

cat(sprintf("Wrote: %s\n", normalizePath(output_file)))
